common-npm-packages/webdeployment-common/msdeployutility.ts

```diff
@@ -206,82 +207,156 @@ function parseAdditionalArguments(additionalArguments: string): string[] {
+                if (!isNaN(subKeyVersion)) {
+                    var subKeyVersionNumber = parseFloat(subKeyVersion);
+                    if (subKeyVersionNumber > latestKeyVersion) {
+                        latestKeyVersion = subKeyVersionNumber;
+                        latestSubKey = subRegKey;
+                    }
                 }
             }
-        }
-        if(latestKeyVersion < 3) {
-            defer.reject(tl.loc("UnsupportedinstalledversionfoundforMSDeployversionshouldbeatleast3orabove", latestKeyVersion));
-            return;
-        }
-        defer.resolve(latestSubKey);
+            if (latestKeyVersion < 3) {
+                reject(tl.loc("UnsupportedinstalledversionfoundforMSDeployversionshouldbeatleast3orabove", latestKeyVersion));
+                return;
+            }
+            resolve(latestSubKey);
+        });
     });
-    return defer.promise;
 }
 
-function getMSDeployInstallPath(registryKey: string): Q.Promise<string> {
-    var defer = Q.defer<string>();
+function getMSDeployInstallPath(registryKey: string): Promise<string> {
+    return new Promise<string>((resolve, reject) => {
+        var regKey = new winreg({
+            hive: winreg.HKLM,
+            key: registryKey
+        })
 
-    var regKey = new winreg({
-      hive: winreg.HKLM,
-      key:  registryKey
-    })
+        regKey.values(function (err, items: { name: string, value: string }[]) {
+            if (err) {
+                reject(tl.loc("UnabletofindthelocationofMSDeployfromregistryonmachineError", err));
+            }
 
-    regKey.get("InstallPath", function(err,item) {
-        if(err) {
-            defer.reject(tl.loc("UnabletofindthelocationofMSDeployfromregistryonmachineError", err));
-            return;
-        }
-        defer.resolve(item.value);
+            if (shouldUseMSDeployTokenAuth()) {
+                const versionItem = items.find(item => item.name === "Version");
+                if (!versionItem) {
+                    reject(tl.loc("MissingMSDeployVersionRegistryKey"));
+                }
+
+                const minimalSupportedVersion = "9.0.7225.0";
+                const version = versionItem.value;
+                tl.debug(`Installed MSDeploy Version: ${version}`);
+
+                // MSDeploy 9.0.7225.0 is the first version to support token auth
+                if (compareVersions(version, minimalSupportedVersion) < 0) {
+                    reject(tl.loc("UnsupportedMSDeployVersion", version));
+                }
+            }
+
+            const installPathItem = items.find(item => item.name === "InstallPath");
+            if (!installPathItem) {
+                reject(tl.loc("MissingMSDeployInstallPathRegistryKey"));
+            }
+
+            resolve(installPathItem.value);
+        });
     });
+}
+
+function compareVersions(version1: string, version2: string): number {
```

Comment:
Can we replace this method with https://github.com/semver/semver ?
Like this:
```js
import * as semver from 'semver';

const result = semver.compare('1.2.3', '1.2.4');
```

Reply:
Yeah, I was considering that. But it would require to add additional reference to that package. Since we really need basic versions comparison I decided to do it directly by our code.

Reply:
You're right